packages/core/src/api/blockManipulation/commands/moveBlocks/moveBlocks.ts

```diff
@@ -0,0 +1,309 @@
+// done by first saving the offsets of the anchor and head from the before
+// positions of their surrounding blocks, as well as the IDs of those blocks. We
+// can then recreate the selection by finding the blocks with those IDs, getting
+// their before positions, and adding the offsets to those positions.
+function updateBlockSelectionFromData(
+  editor: BlockNoteEditor<any, any, any>,
+  data: BlockSelectionData
+) {
+  const anchorBlockPos = getNodeById(
+    data.anchorBlockId,
+    editor._tiptapEditor.state.doc
+  ).posBeforeNode;
+
+  let selection: Selection;
+  if (data.type === "cell") {
+    selection = CellSelection.create(
+      editor._tiptapEditor.state.doc,
+      anchorBlockPos + data.anchorCellOffset,
+      anchorBlockPos + data.headCellOffset
+    );
+  } else if (data.type === "node") {
+    selection = NodeSelection.create(
+      editor._tiptapEditor.state.doc,
+      anchorBlockPos + 1
+    );
+  } else {
+    const headBlockPos = getNodeById(
+      data.headBlockId,
+      editor._tiptapEditor.state.doc
+    ).posBeforeNode;
+
+    selection = TextSelection.create(
+      editor._tiptapEditor.state.doc,
+      anchorBlockPos + data.anchorOffset,
+      headBlockPos + data.headOffset
+    );
+  }
+
+  editor._tiptapEditor.view.dispatch(
+    editor._tiptapEditor.state.tr.setSelection(selection)
+  );
+}
+
+// Replaces any `columnList` blocks with the children of their columns. This is
+// done here instead of in `getSelection` as we still need to remove the entire
+// `columnList` node but only insert the `blockContainer` nodes inside it.
+function flattenColumns(
+  blocks: Block<any, any, any>[]
+): Block<any, any, any>[] {
+  return blocks
+    .map((block) => {
+      if (block.type === "columnList") {
+        return block.children
+          .map((column) => flattenColumns(column.children))
+          .flat();
+      }
+
+      return {
+        ...block,
+        children: flattenColumns(block.children),
+      };
+    })
+    .flat();
+}
+
+// Removes the selected blocks from the editor, then inserts them before/after a
+// reference block. Also updates the selection to match the original selection
+// using `getBlockSelectionData` and `updateBlockSelectionFromData`.
+export function moveSelectedBlocksAndSelection(
+  editor: BlockNoteEditor<any, any, any>,
+  referenceBlock: BlockIdentifier,
+  placement: "before" | "after"
+) {
+  const blocks = editor.getSelection()?.blocks || [
+    editor.getTextCursorPosition().block,
+  ];
+  const selectionData = getBlockSelectionData(editor);
+
+  editor.removeBlocks(blocks);
+  editor.insertBlocks(flattenColumns(blocks), referenceBlock, placement);
+
+  updateBlockSelectionFromData(editor, selectionData);
+}
+
+// Checks if a block is in a valid place after being moved. This check is
+// primitive at the moment and only returns false if the block's parent is a
+// `columnList` block. This is because when moving blocks, columns are flattened
```

Comment:
> This is because when moving blocks, columns are flattened
// into regular blocks and therefore can only be nested inside `column` or
// other regular blocks

I don't think this is very related to flattening. Imo the main reason (a block can't be a direct child of a columnlist) is enough